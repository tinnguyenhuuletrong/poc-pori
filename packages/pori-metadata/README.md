# pori-metadata

This library was generated with [Nx](https://nx.dev).

## Building

Run `nx build pori-metadata` to build the library.

## Running unit tests

Run `nx test pori-metadata` to execute the unit tests via [Jest](https://jestjs.io).
