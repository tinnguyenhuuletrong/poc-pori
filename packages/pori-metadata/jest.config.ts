export default {
  displayName: 'pori-metadata',
  preset: '../../jest.preset.js',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    },
  },
  transform: {
    '^.+\\.[tj]s$': 'ts-jest',
  },
  moduleFileExtensions: ['ts', 'js', 'html', 'json'],
  coverageDirectory: '../../coverage/packages/pori-metadata',
};
