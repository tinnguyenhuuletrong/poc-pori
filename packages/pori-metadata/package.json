{
  "name": "@pori-and-friends/pori-metadata",
  "version": "0.0.1",
  "type": "commonjs"
}
