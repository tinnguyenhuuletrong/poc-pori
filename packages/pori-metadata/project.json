{
  "sourceRoot": "packages/pori-metadata/src",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/pori-metadata",
        "main": "packages/pori-metadata/src/index.ts",
        "tsConfig": "packages/pori-metadata/tsconfig.lib.json",
        "assets": ["packages/pori-metadata/*.md"]
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/pori-metadata/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/packages/pori-metadata"],
      "options": {
        "jestConfig": "packages/pori-metadata/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
