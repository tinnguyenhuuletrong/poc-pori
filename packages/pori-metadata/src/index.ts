import { Context, CustomEstGasprice, ENV } from './commonTypes';
import * as stagConfig from './lib/sta-poriverse_info';
import * as prodConfig from './lib/prod-poriverse_info';
import * as prodPoriChainConfig from './lib/prod-porichain-poriverse_info';
import { mean } from 'lodash';

export const TEN_POWER_10_BN = BigInt(10 ** 18);
export const TURN_DURATION_SEC = 1800;

export function getWeb3NodeUri(env: ENV) {
  const key = `NODE_URI_${env}`.toUpperCase();
  return process.env[key] as string;
}

export function getWeb3NodeUriHttp(env: ENV) {
  const key = `NODE_URI_${env}_HTTP`.toUpperCase();
  return process.env[key] as string;
}

export function getWeb3NodeUriPolygonHttp() {
  const key = `NODE_URI_${ENV.Prod}_HTTP`.toUpperCase();
  return process.env[key] as string;
}

export function getAPILink(env: ENV) {
  if (env === ENV.Staging) return stagConfig.gameInfo.m.app.apiUrl;
  else if (env === ENV.ProdPorichain)
    return prodPoriChainConfig.gameInfo.m.app.apiUrl;
  else return prodConfig.gameInfo.m.app.apiUrl;
}

export function getKyberSwapFactoryAddress(env: ENV) {
  // https://docs.kyberswap.com/developer-guides/kyberswap-addresses#mainnet

  if (env === ENV.Staging) return '0x7900309d0b1c8D3d665Ae40e712E8ba4FC4F5453';
  return '0x5f1fe642060b5b9658c15721ea22e982643c095c';
}

export function getRIGYTokenInfoOnPolygon() {
  const tokenConfig = prodConfig.gameInfo.m.app;
  return {
    symbol: tokenConfig.token.inGameSymbol,
    tokenAddress: tokenConfig.token.inGameAddress,
    decimal: tokenConfig.token.inGameDecimal,
    chainId: tokenConfig.rpcMetamask.chainId,
  };
}

export function getRIGYTokenInfo(env: ENV) {
  let tokenConfig = prodConfig.gameInfo.m.app;
  if (env === ENV.Staging) {
    tokenConfig = stagConfig.gameInfo.m.app;
  } else if (env === ENV.ProdPorichain)
    tokenConfig = prodPoriChainConfig.gameInfo.m.app;

  return {
    symbol: tokenConfig.token.inGameSymbol,
    tokenAddress: tokenConfig.token.inGameAddress,
    decimal: tokenConfig.token.inGameDecimal,
    chainId: tokenConfig.rpcMetamask.chainId,
  };
}

export function getRIKENTokenInfoOnPolygon() {
  const tokenConfig = prodConfig.gameInfo.m.app;
  return {
    symbol: tokenConfig.token.nativeSymbol,
    tokenAddress: tokenConfig.token.nativeAddress,
    decimal: tokenConfig.token.nativeDecimal,
    chainId: tokenConfig.rpcMetamask.chainId,
  };
}

export function getRIKENTokenInfo(env: ENV) {
  let tokenConfig = prodConfig.gameInfo.m.app;
  if (env === ENV.Staging) {
    tokenConfig = stagConfig.gameInfo.m.app;
  } else if (env === ENV.ProdPorichain)
    tokenConfig = prodPoriChainConfig.gameInfo.m.app;

  return {
    symbol: tokenConfig.token.nativeSymbol,
    tokenAddress: tokenConfig.token.nativeAddress,
    decimal: tokenConfig.token.nativeDecimal,
    chainId: tokenConfig.rpcMetamask.chainId,
  };
}

export function getAdventureBaseLink(env: ENV) {
  if (env === ENV.Staging) return stagConfig.gameInfo.m.app.urlAdventure;
  else if (env === ENV.ProdPorichain)
    return prodPoriChainConfig.gameInfo.m.app.urlAdventure;
  return prodConfig.gameInfo.m.app.urlAdventure;
}

export function getMarketplayBaseLink(env: ENV) {
  if (env === ENV.Staging) return stagConfig.gameInfo.m.app.urlMarketplace;
  else if (env === ENV.ProdPorichain)
    return prodPoriChainConfig.gameInfo.m.app.urlMarketplace;
  return prodConfig.gameInfo.m.app.urlMarketplace;
}

export function getIdleGameAddressSC(env: ENV) {
  if (env === ENV.Staging) {
    return {
      abi: stagConfig.ABI_IDLE,
      address: stagConfig.gameInfo.m.app.contractAddress.idleGameAddress,
      createdBlock: stagConfig.gameInfo.m.app.scCreatedBlock.idle,
    };
  } else if (env === ENV.ProdPorichain)
    return {
      abi: prodPoriChainConfig.ABI_IDLE,
      address:
        prodPoriChainConfig.gameInfo.m.app.contractAddress.idleGameAddress,
      createdBlock: prodPoriChainConfig.gameInfo.m.app.scCreatedBlock.idle,
    };
  return {
    abi: prodConfig.ABI_IDLE,
    address: prodConfig.gameInfo.m.app.contractAddress.idleGameAddress,
    createdBlock: prodConfig.gameInfo.m.app.scCreatedBlock.idle,
  };
}

export function calculateMineTurnTime(startTime: Date) {
  const farmerAtkStartAt = new Date(
    startTime.valueOf() + TURN_DURATION_SEC * 1000
  );
  const supporterAtkStartAt = new Date(
    startTime.valueOf() + TURN_DURATION_SEC * 2 * 1000
  );
  return {
    farmerAtkStartAt,
    supporterAtkStartAt,
  };
}

export function getContextSetting(env: ENV) {
  // getGas porichain return as 500. But avg fee on chain around 4100
  const gasFactor = 1;
  const safeGweith = env === ENV.ProdPorichain ? 5000 : 80;
  const autoPlayMicroDelayMs = env === ENV.ProdPorichain ? 10000 : 3000;

  // custom estimage gas function base on pending tx for porichain
  let estimageGas: CustomEstGasprice | undefined;
  if (env === ENV.ProdPorichain) {
    estimageGas = async (ctx: Context) => {
      const pendingTx = await ctx.web3.eth.getPendingTransactions();
      const avgGas = mean(
        pendingTx
          .filter((itm) => +itm.gasPrice > 0)
          .map((itm) => {
            return +itm.gasPrice;
          })
      );
      return avgGas.toString();
    };
  }

  return {
    setting: { gasFactor, safeGweith, autoPlayMicroDelayMs },
    custom: {
      estimageGas,
    },
  };
}

export * from './commonTypes';

export * as IdleGameSc from './lib/idleGameSc';
export * from './lib/idleGameSc/type.idleGame';
