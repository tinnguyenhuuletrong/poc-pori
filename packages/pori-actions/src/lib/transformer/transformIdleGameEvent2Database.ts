import Realm from 'realm';
import * as Repos from '@pori-and-friends/pori-repositories';
import { IdleGameSc, NftInfo } from '@pori-and-friends/pori-metadata';

export async function transformIdleGameEvent2Database(
  realm: Realm,
  events: Repos.IdleGameSCEventDataModel[],
  resolveNftInfo: (id: string | number) => Promise<NftInfo>,
  onIt?: (it: Realm.BSON.ObjectId) => void
) {
  for (const it of events) {
    switch (it.type) {
      case IdleGameSc.EIdleGameSCEventType.PorianDeposited:
        {
          const data = it.data as IdleGameSc.PorianDepositedData;
          const playerId = data.from;
          const poriId = data.porian;

          let nftInfo: any;

          const poriObj = await Repos.PoriRepo.findOne(realm, poriId);

          // resolve api query only for new nft
          if (!poriObj) nftInfo = await resolveNftInfo(poriId);

          Repos.PlayerRepo.txSync(realm, () => {
            const playerObj = Repos.PlayerRepo.getOrCreate(
              realm,
              playerId,
              Repos.PlayerDataModel.generate(playerId, it.blockNo)
            );

            const poriObj = Repos.PoriRepo.getOrCreate(
              realm,
              poriId,
              nftInfo ? Repos.PoriDataModel.generate(nftInfo) : {}
            );

            // active + add to player inventories
            poriObj.isActive = true;
            const isExists =
              playerObj.pories.findIndex((itm) => itm._id === poriId) >= 0;
            if (!isExists) playerObj.pories.push(poriObj);
          });
        }
        break;

      case IdleGameSc.EIdleGameSCEventType.PorianWithdrawed:
        {
          const data = it.data as IdleGameSc.PorianWithdrawedData;
          const playerId = data.to;
          const poriId = data.porian;

          const playerObj = await Repos.PlayerRepo.findOne(realm, playerId);
          const poriObj = await Repos.PoriRepo.findOne(realm, poriId);

          if (!playerObj) break;

          Repos.PlayerRepo.txSync(realm, () => {
            // deactive + remove from player inventories
            if (poriObj) poriObj.isActive = false;
            const isExists =
              playerObj.pories.findIndex((itm) => itm._id === poriId) >= 0;
            if (isExists) {
              playerObj.pories = playerObj.pories.filter(
                (itm) => itm._id !== poriId
              ) as any;
            }
          });
        }
        break;

      case IdleGameSc.EIdleGameSCEventType.GameDurationChanged:
        {
          const data = it.data as IdleGameSc.GameDurationChangedData;
          const metadata = await Repos.IdleGameSCMetadataRepo.findOne(
            realm,
            'default'
          );

          if (!metadata) break;

          Repos.IdleGameSCMetadataRepo.txSync(realm, () => {
            metadata.extras['turnDuration'] = data.turnDuration;
            metadata.extras['adventureDuration'] = data.adventureDuration;
          });
        }
        break;

      default:
        break;
    }
    onIt && onIt(it._id);
  }
}
