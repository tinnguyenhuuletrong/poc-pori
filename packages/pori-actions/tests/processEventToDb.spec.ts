import { IdleGameSc } from '@pori-and-friends/pori-metadata';
import {
  openRepo,
  IdleGameSCEventDataModel,
  PlayerRepo,
  PoriRepo,
} from '@pori-and-friends/pori-repositories';
import { transformIdleGameEvent2Database } from '../src/lib/transformer/transformIdleGameEvent2Database';

describe('processEventToDb', () => {
  let realm: Realm;

  beforeAll(async () => {
    realm = await openRepo({
      path: './_tests/testRepo.realm',
      inMemory: true,
    });
  });

  afterEach(() => {
    realm.write(() => {
      realm.deleteAll();
    });
  });

  afterAll(() => {
    realm.close();
  });

  test('dataset_1', async () => {
    const Data = await import('./_mocks/dataset_1');
    const res = await transformIdleGameEvent2Database(
      realm,
      Data.EVENTS,
      Data.resolveNftInfo
    );

    const players = (await PlayerRepo.findAll(realm)).toJSON();
    const pories = (await PoriRepo.findAll(realm)).toJSON();

    expect({
      players,
      pories,
    }).toMatchSnapshot();
  });
});
