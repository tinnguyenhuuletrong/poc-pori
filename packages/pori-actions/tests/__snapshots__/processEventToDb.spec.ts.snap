// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processEventToDb dataset_1 1`] = `
Object {
  "players": Array [
    Object {
      "_id": "0x3048aa7a4eb54bdc7f7b35d98603a18718380993",
      "createdBlock": 1,
      "pories": Array [
        Object {
          "_id": 1182,
          "birthDate": "2022-04-20T11:28:00.000Z",
          "createdAt": "2022-04-20T12:53:03.401Z",
          "critical": 71,
          "dna": "9076431867127095074785173926980014995178110010806707409896581273537522696192",
          "health": 86,
          "helpPower": 294,
          "isActive": true,
          "legend": 0,
          "magicAttack": 69,
          "magicDefend": 74,
          "minePower": 310,
          "name": "Bruwan",
          "ownerAddress": "0x3048aa7a4eb54bdc7f7b35d98603a18718380993",
          "physicalAttack": 69,
          "physicalDefend": 79,
          "player": Array [
            [Circular],
          ],
          "poriClass": 1,
          "purity": 3,
          "speed": 70,
          "stage": 2,
          "status": 5,
          "type": 2,
          "updatedAt": "2022-04-27T15:11:03.202Z",
        },
      ],
    },
  ],
  "pories": Array [
    Object {
      "_id": 1182,
      "birthDate": "2022-04-20T11:28:00.000Z",
      "createdAt": "2022-04-20T12:53:03.401Z",
      "critical": 71,
      "dna": "9076431867127095074785173926980014995178110010806707409896581273537522696192",
      "health": 86,
      "helpPower": 294,
      "isActive": true,
      "legend": 0,
      "magicAttack": 69,
      "magicDefend": 74,
      "minePower": 310,
      "name": "Bruwan",
      "ownerAddress": "0x3048aa7a4eb54bdc7f7b35d98603a18718380993",
      "physicalAttack": 69,
      "physicalDefend": 79,
      "player": Array [
        Object {
          "_id": "0x3048aa7a4eb54bdc7f7b35d98603a18718380993",
          "createdBlock": 1,
          "pories": Array [
            [Circular],
          ],
        },
      ],
      "poriClass": 1,
      "purity": 3,
      "speed": 70,
      "stage": 2,
      "status": 5,
      "type": 2,
      "updatedAt": "2022-04-27T15:11:03.202Z",
    },
    Object {
      "_id": 1183,
      "birthDate": "2022-04-20T11:28:00.000Z",
      "createdAt": "2022-04-20T12:53:03.401Z",
      "critical": 71,
      "dna": "9076431867127095074785173926980014995178110010806707409896581273537522696192",
      "health": 86,
      "helpPower": 294,
      "isActive": false,
      "legend": 0,
      "magicAttack": 69,
      "magicDefend": 74,
      "minePower": 310,
      "name": "Bruwan",
      "ownerAddress": "0x3048aa7a4eb54bdc7f7b35d98603a18718380993",
      "physicalAttack": 69,
      "physicalDefend": 79,
      "player": Array [],
      "poriClass": 1,
      "purity": 3,
      "speed": 70,
      "stage": 2,
      "status": 5,
      "type": 2,
      "updatedAt": "2022-04-27T15:11:03.202Z",
    },
  ],
}
`;
